File: apps/worker/adapters/rapidapi_flights.py
# apps/worker/adapters/rapidapi_flights.py
#
# Sky Scrapper API via RapidAPI — capped at 10 calls per calendar month.
# Each call targets the midpoint date of the window to preserve quota.

import os
import requests
from datetime import date

import db  # local db.py

RAPID_API_HOST = "sky-scrapper.p.rapidapi.com"
MONTHLY_LIMIT = 10


def _can_call() -> bool:
    year_month = date.today().strftime("%Y-%m")
    used = db.get_rapidapi_usage(year_month)
    return used < MONTHLY_LIMIT


def _format_duration(minutes: int) -> str:
    h, m = divmod(minutes, 60)
    return f"{h}h {m}m"


def _midpoint_date(date_from: str, date_to: str) -> str:
    d1 = date.fromisoformat(date_from)
    d2 = date.fromisoformat(date_to)
    mid = d1 + (d2 - d1) / 2
    return mid.isoformat()


def scrape_cash_midpoint(monitor: dict, api_key: str) -> list[dict]:
    """
    Make ONE RapidAPI call for the midpoint date of the window.
    Returns up to 5 results or [] if limit reached / error.
    """
    if not _can_call():
        print(f"  [RapidAPI] Monthly limit of {MONTHLY_LIMIT} reached — skipping")
        return []

    origin = monitor["origin"]
    destination = monitor["destination"]
    mid_date = _midpoint_date(monitor["date_from"], monitor["date_to"])
    monitor_id = monitor["id"]

    url = (
        f"https://{RAPID_API_HOST}/api/v1/flights/searchFlights"
        f"?originSkyId={origin}&destinationSkyId={destination}"
        f"&originEntityId=&destinationEntityId="
        f"&date={mid_date}&returnDate=&cabinClass=economy&adults=1"
        f"&sortBy=best&currency=CAD&market=en-CA&countryCode=CA"
    )
    headers = {
        "x-rapidapi-host": RAPID_API_HOST,
        "x-rapidapi-key": api_key,
    }

    try:
        print(f"  [RapidAPI] Calling {origin}→{destination} on {mid_date}")
        resp = requests.get(url, headers=headers, timeout=20)
        resp.raise_for_status()
        data = resp.json()

        year_month = date.today().strftime("%Y-%m")
        db.increment_rapidapi_usage(year_month)

        itineraries = data.get("data", {}).get("itineraries", [])
        results = []

        for it in itineraries[:5]:
            price_raw = it.get("price", {}).get("raw", 0)
            legs = it.get("legs", [])
            leg = legs[0] if legs else {}
            carrier = (leg.get("carriers", {}).get("marketing") or [{}])[0]

            results.append({
                "provider": "RapidAPI (Sky Scrapper)",
                "monitor_id": monitor_id,
                "kind": "cash",
                "origin": origin,
                "destination": destination,
                "departure_date": mid_date,
                "total_price": round(float(price_raw), 2),
                "currency": "CAD",
                "airline": carrier.get("name", "Unknown"),
                "flight_number": carrier.get("alternateId", "") + str(leg.get("flightNumber", "")),
                "stops": leg.get("stopCount", 0),
                "duration": _format_duration(leg.get("durationInMinutes", 0)),
                "booking_url": (
                    f"https://www.google.com/travel/flights?q=Flights+from+"
                    f"{origin}+to+{destination}+on+{mid_date}"
                ),
                "is_award": False,
            })

        print(f"  [RapidAPI] Got {len(results)} results")
        return results

    except Exception as e:
        print(f"  [RapidAPI] Error: {e}")
        return []

File: apps/worker/src/adapters/rapidapi-flights.ts
// apps/worker/src/adapters/rapidapi-flights.ts
//
// Uses the "Sky Scrapper" API on RapidAPI (host: sky-scrapper.p.rapidapi.com)
// which provides Google Flights data via official-ish API.
//
// IMPORTANT: Limited to 10 calls per calendar month. Usage is tracked in
// the `rapidapi_usage` database table. Once the limit is reached this
// adapter returns empty results so the Playwright adapter takes over.

import { CashMonitor, AwardMonitor, FlightResult } from "@flight-tracker/shared";
import { ScrapeAdapter } from "./base";
import { db } from "../db/client";

const RAPID_API_HOST = "sky-scrapper.p.rapidapi.com";
const MONTHLY_LIMIT = 10;

interface SkyScrapeFlightLeg {
  origin: { displayCode: string };
  destination: { displayCode: string };
  departure: string;
  arrival: string;
  durationInMinutes: number;
  carriers: { marketing: Array<{ name: string; alternateId: string }> };
  flightNumber: string;
  stopCount: number;
}

interface SkyScrapeItinerary {
  price: { raw: number; formatted: string };
  legs: SkyScrapeFlightLeg[];
}

export class RapidApiFlightsAdapter implements ScrapeAdapter {
  name = "RapidAPI (Sky Scrapper)";
  private apiKey: string;

  constructor(apiKey: string) {
    this.apiKey = apiKey;
  }

  /** Returns how many RapidAPI calls have been made this calendar month */
  private async getMonthlyUsage(): Promise<number> {
    const yearMonth = new Date().toISOString().slice(0, 7); // YYYY-MM
    try {
      const row = await db.query(
        "SELECT call_count FROM rapidapi_usage WHERE year_month = $1",
        [yearMonth]
      );
      return row.rows[0]?.call_count ?? 0;
    } catch {
      return 0;
    }
  }

  /** Increments the monthly usage counter */
  private async incrementUsage(): Promise<void> {
    const yearMonth = new Date().toISOString().slice(0, 7);
    await db.query(
      `INSERT INTO rapidapi_usage (year_month, call_count)
       VALUES ($1, 1)
       ON CONFLICT (year_month)
       DO UPDATE SET call_count = rapidapi_usage.call_count + 1`,
      [yearMonth]
    );
  }

  /** Check if we can make another call this month */
  async canCallThisMonth(): Promise<boolean> {
    const used = await this.getMonthlyUsage();
    return used < MONTHLY_LIMIT;
  }

  private formatDuration(minutes: number): string {
    const h = Math.floor(minutes / 60);
    const m = minutes % 60;
    return `${h}h ${m}m`;
  }

  private async fetchFlights(
    origin: string,
    destination: string,
    date: string, // YYYY-MM-DD
    monitorId: string
  ): Promise<FlightResult[]> {
    const url =
      `https://${RAPID_API_HOST}/api/v1/flights/searchFlights?` +
      `originSkyId=${origin}&destinationSkyId=${destination}` +
      `&originEntityId=&destinationEntityId=` +
      `&date=${date}&returnDate=&cabinClass=economy&adults=1` +
      `&sortBy=best&currency=CAD&market=en-CA&countryCode=CA`;

    const resp = await fetch(url, {
      headers: {
        "x-rapidapi-host": RAPID_API_HOST,
        "x-rapidapi-key": this.apiKey,
      },
    });

    if (!resp.ok) {
      throw new Error(`RapidAPI HTTP ${resp.status}: ${await resp.text()}`);
    }

    const data = await resp.json();
    const itineraries: SkyScrapeItinerary[] =
      data?.data?.itineraries ?? [];

    return itineraries.slice(0, 5).map((it) => {
      const leg = it.legs[0];
      const carrier = leg?.carriers?.marketing?.[0];
      return {
        provider: this.name,
        monitorId,
        origin,
        destination,
        departureDate: date,
        totalPrice: it.price.raw,
        currency: "CAD",
        airline: carrier?.name ?? "Unknown",
        flightNumber: carrier?.alternateId
          ? `${carrier.alternateId}${leg?.flightNumber ?? ""}`
          : leg?.flightNumber ?? "",
        stops: leg?.stopCount ?? 0,
        duration: this.formatDuration(leg?.durationInMinutes ?? 0),
        bookingUrl: `https://www.google.com/travel/flights?q=Flights+from+${origin}+to+${destination}+on+${date}`,
        scrapedAt: new Date().toISOString(),
        isAward: false,
      };
    });
  }

  async scrapeWindow(monitor: CashMonitor): Promise<FlightResult[]> {
    if (!(await this.canCallThisMonth())) {
      console.log(
        `[RapidAPI] Monthly limit of ${MONTHLY_LIMIT} reached — skipping`
      );
      return [];
    }

    const results: FlightResult[] = [];

    // For window monitors use ONE call on the midpoint date to save quota.
    // Playwright handles the full window; RapidAPI just supplements with
    // a single targeted call.
    const start = new Date(monitor.dateFrom);
    const end = new Date(monitor.dateTo);
    const midMs = (start.getTime() + end.getTime()) / 2;
    const midDate = new Date(midMs).toISOString().split("T")[0];

    try {
      console.log(
        `[RapidAPI] Calling for midpoint ${monitor.origin}→${monitor.destination} on ${midDate}`
      );
      const flights = await this.fetchFlights(
        monitor.origin,
        monitor.destination,
        midDate,
        monitor.id
      );
      await this.incrementUsage();
      results.push(...flights);
    } catch (err) {
      console.error(`[RapidAPI] Error:`, err);
    }

    return results;
  }

  async scrapeMonth(_monitor: AwardMonitor): Promise<FlightResult[]> {
    // RapidAPI cash-fare search doesn't support award availability
    return [];
  }

  async isHealthy(): Promise<boolean> {
    return this.apiKey?.length > 0 && (await this.canCallThisMonth());
  }

  async close(): Promise<void> {
    // no resources to free
  }
}
